### PR TITLE
⛽ Reduce gas cost of sqrt using ABDK Consulting implementation

### PR DESCRIPTION
**Pull request by rajivpo**
The ABDK Consulting [`sqrt` function implementation](https://github.com/abdk-consulting/abdk-libraries-solidity/blob/master/ABDKMath64x64.sol) often requires less the gas than the current `sqrt` implementation.

As a concrete example I implemented a temporary test function (not included in the PR).

```
function testSqrtSingle() public {
        // 99972 was chosen arbitrarily
        assertEq(FixedPointMath.sqrt(9994400784), 99972);
}
```

Using the current implementation, I got the following: 
image

Using the proposed implementation without it being wrapped with `unchecked`, I got the following:
image

Using the proposed implemented wrapped with `unchecked, I got the following: 
image

For the sake of testing diversity here are the outputs for a few additional gas comparisons:

```
[PASS] testSqrt_3_Current() (gas: 354)
[PASS] testSqrt_3_Proposed_Unchecked() (gas: 867)
[PASS] testSqrt_16_Current() (gas: 1238)
[PASS] testSqrt_16_Proposed_Unchecked() (gas: 911)
[PASS] testSqrt_2704_Current() (gas: 2201)
[PASS] testSqrt_2704_Proposed_Unchecked() (gas: 895)
[PASS] testSqrt_110889_Current() (gas: 3205)
[PASS] testSqrt_110889_Proposed_Unchecked() (gas: 890)
[PASS] testSqrt_32239684_Current() (gas: 4146)
[PASS] testSqrt_32239684_Proposed_Unchecked() (gas: 980)
[PASS] testSqrt_152399025_Current() (gas: 4414)
[PASS] testSqrt_152399025_Proposed_Unchecked() (gas: 942)
```

I think this implementation is worse for some (e.g 3) but not most numbers it appears.

@TransmissionsDev found something similar with what I imagine is a similar test

image

In terms of correctness of this implementation compared to the current, we have the following (link to [tweet](https://twitter.com/sendmoodz/status/1428702803149656067)):
image

### CODE REVIEW

**Comment by TransmissionsDev** [comment]
we love a king who conforms to my emoji standards

File: src/utils/FixedPointMath.sol [hunk 1]
```diff
@@ -94,16 +94,28 @@ library FixedPointMath {
...
+            if (xx >= 0x8) { r <<= 1; }
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1; // Seven iterations should be enough
+            uint256 r1 = x / r;
+            return uint128 (r < r1 ? r : r1);
```
Comment (TransmissionsDev) [inline src/utils/FixedPointMath.sol:118]: do we need this?

File: src/utils/FixedPointMath.sol [hunk 2]
```diff
@@ -94,16 +94,28 @@ library FixedPointMath {
...
+            if (xx >= 0x100) { xx >>= 8; r <<= 4; }
+            if (xx >= 0x10) { xx >>= 4; r <<= 2; }
+            if (xx >= 0x8) { r <<= 1; }
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1;
+            r = (r + x / r) >> 1; // Seven iterations should be enough
```
Comment (TransmissionsDev) [inline src/utils/FixedPointMath.sol:116]: lets remove the creepy seven iterations thing lol

File: src/utils/FixedPointMath.sol [hunk 3]
```diff
@@ -94,16 +94,28 @@ library FixedPointMath {
...
-        if (y > 3) {
-            z = y;
-            uint256 x = y / 2 + 1;
-            while (x < z) {
-                z = x;
-                x = (y / x + x) / 2;
-            }
-        } else if (y != 0) {
-            z = 1;
+    // Credit to ABDK Consulting (https://github.com/abdk-consulting/abdk-libraries-solidity/blob/master/ABDKMath64x64.sol)
```
Comment (TransmissionsDev) [inline src/utils/FixedPointMath.sol:97]: Can we remove this, credit at the top is fine imo

File: src/utils/FixedPointMath.sol [hunk 4]
```diff
@@ -2,7 +2,7 @@
 pragma solidity >=0.8.0;
 
 /// @notice Arithmetic library with operations for fixed-point numbers.
-/// @author Modified from Dappsys V2 (https://github.com/dapp-org/dappsys-v2/blob/main/src/math.sol)
+/// @author Modified from Dappsys V2 (https://github.com/dapp-org/dappsys-v2/blob/main/src/math.sol) and ABDK Consulting (https://github.com/abdk-consulting/abdk-libraries-solidity/blob/master/ABDKMath64x64.sol)
```
Comment (TransmissionsDev) [inline src/utils/FixedPointMath.sol:5]: i am OCD about comment alignment lol

can we make this two lines? keep the first line as it was before and make the second

```
/// and ABDK (https://github.com/abdk-consulting/abdk-libraries-solidity/blob/master/ABDKMath64x64.sol)
```

File: src/utils/FixedPointMath.sol [hunk 5]
```diff
@@ -94,16 +94,28 @@ library FixedPointMath {
...
-            z = y;
-            uint256 x = y / 2 + 1;
-            while (x < z) {
-                z = x;
-                x = (y / x + x) / 2;
-            }
-        } else if (y != 0) {
-            z = 1;
+    // Credit to ABDK Consulting (https://github.com/abdk-consulting/abdk-libraries-solidity/blob/master/ABDKMath64x64.sol)
+    function sqrt(uint256 x) internal pure returns (uint256 z) {
```
Comment (TransmissionsDev) [inline src/utils/FixedPointMath.sol:98]: isn't all this logic wrapped in an unchecked block in ABDK's impl?

**Review by TransmissionsDev — CHANGES_REQUESTED** [review]

File: src/utils/FixedPointMath.sol (hunk 5 above)
Comment (rajivpo) [reply src/utils/FixedPointMath.sol:98]: Not in the ABDK impl but it is in the HiFi-Finance impl: https://github.com/hifi-finance/prb-math/blob/main/contracts/PRBMath.sol

Comment (rajivpo) [reply src/utils/FixedPointMath.sol:98]: nope jk it is in the ABDK impl: https://github.com/abdk-consulting/abdk-libraries-solidity/blob/master/ABDKMath64x64.sol

will update

File: src/utils/FixedPointMath.sol (hunk 1 above)
Comment (rajivpo) [reply src/utils/FixedPointMath.sol]: nah as per what you surfaced via https://github.com/Uniswap/uniswap-lib/blob/c01640b0f0f1d8a85cba8de378cc48469fcfd9a6/contracts/libraries/Babylonian.sol#L10

**Comment by rajivpo** [comment]
Locally, wrote this: 

```
function testSqrtDiff(uint256 x) public {
        uint256 rootCurrent = FixedPointMath.sqrtOld(x);
        uint256 rootProposed = FixedPointMath.sqrt(x);
        uint256 next = rootCurrent + 1;

        // Ignore cases where next * next overflows.
        unchecked {
            if (next * next < next) return;
        }

        assertTrue(rootCurrent * rootCurrent <= x && rootProposed * rootProposed <= x && rootProposed *rootProposed == rootCurrent * rootCurrent && next * next > x);
    }
```

and fuzzed for 500k runs resulting in this

image


Will evaluate again tm, but possibly bullish for abdk implementation?

**Comment by rajivpo** [comment]
I've also added a `make lint` command to make development easier for future contributors. If preferred, I'm happy to move it into a separate PR.

**Comment by TransmissionsDev** [comment]
Have we compared this ABDK impl gas usage with the current impl but with an unchecked wrapper?

**Comment by rajivpo** [comment]
> Have we compared this ABDK impl gas usage with the current impl but with an unchecked wrapper?

We have now.

```
[PASS] testSqrt_3_Current_Unchecked() (gas: 309)
[PASS] testSqrt_3_Proposed_Unchecked() (gas: 861)
[PASS] testSqrt_16_Current_Unchecked() (gas: 684)
[PASS] testSqrt_16_Proposed_Unchecked() (gas: 905)
[PASS] testSqrt_2704_Current_Unchecked() (gas: 1141)
[PASS] testSqrt_2704_Proposed_Unchecked() (gas: 889)
[PASS] testSqrt_110889_Current_Unchecked() (gas: 1556)
[PASS] testSqrt_110889_Proposed_Unchecked() (gas: 929)
[PASS] testSqrt_32239684_Current_Unchecked() (gas: 1993)
[PASS] testSqrt_32239684_Proposed_Unchecked() (gas: 952)
```

Reference implementation of Current with `unchecked`:
```
function sqrtOld(uint256 y) internal pure returns (uint256 z) {
        unchecked {
            if (y > 3) {
                z = y;
                uint256 x = y / 2 + 1;
                while (x < z) {
                    z = x;
                    x = (y / x + x) / 2;
                }
            } else if (y != 0) {
                z = 1;
            }
        }
    }
```

**Comment by TransmissionsDev** [comment]
Interesting look like the current impl is more efficient for low nums but ABDK starts dominating st higher numbers. Considering most ppl are probably dealing with fixed point numbers which will be massive, ABDK seems like the best option 👍

**Review by TransmissionsDev — APPROVED** [review]

**Pull request merged**